File: ifitwala_ed/students/page/student_log_dashboard/student_log_dashboard.js
// Copyright (c) 2025, François de Ryckel and contributors
// For license information, please see license.txt

// Copyright (c) 2025, François de Ryckel and contributors
// For license information, please see license.txt

let selected_student = null;
let recent_start = 0;
const recent_page_len = 25;
let branchSchools = [];

/*── 0.1 recent log table ─────────────────────────────────────────────────*/
function fetch_recent_logs(page, append = false) {
	const fd = page.fields_dict;
	const safe = name =>
		fd && fd[name] && typeof fd[name].get_value === "function"
			? fd[name].get_value() : null;

	frappe.call({
		method: "ifitwala_ed.students.page.student_log_dashboard.student_log_dashboard.get_recent_logs",
		args: {
			filters: JSON.stringify({
				school: safe("school"),
				academic_year: safe("academic_year"),
				program: safe("program"),
				author: safe("author"),
				from_date: safe("from_date"),
				to_date: safe("to_date"),
			}),
			start: recent_start,
			page_length: recent_page_len,
		},
		callback: (r) => {
			if (!r.message || r.message.error) return;

			const rows = r.message.map(d => `
				<tr>
					<td>${d.date}</td>
					<td>${d.student}</td>
					<td>${d.program}</td>
					<td>${d.log_type}</td>
					<td>${d.content}</td>
					<td>${d.author}</td>
					<td class="text-center">${d.requires_follow_up ? "✓" : "✗"}</td>
				</tr>`).join("");

			const body = document.getElementById("recent-log-table-body");
			if (!append) body.innerHTML = "";
			body.insertAdjacentHTML("beforeend", rows);

			if (r.message.length < recent_page_len) {
				document.getElementById("recent-log-load-more").style.display = "none";
			} else {
				document.getElementById("recent-log-load-more").style.display = "block";
			}
			recent_start += r.message.length;
		}
	});
}

/*── 0.2 dashboard data ───────────────────────────────────────────────────*/
function fetch_dashboard_data(page) {
	const fd = page.fields_dict;
	const safe = (name) =>
		fd && fd[name] && typeof fd[name].get_value === "function"
			? fd[name].get_value() : null;

	const filters = {
		school: safe("school"),
		academic_year: safe("academic_year"),
		program: safe("program"),
		student: selected_student,
		author: safe("author"),
		from_date: safe("from_date"),
		to_date: safe("to_date"),
	};

	frappe.call({
		method: "ifitwala_ed.students.page.student_log_dashboard.student_log_dashboard.get_dashboard_data",
		args: { filters },
		callback: (r) => {
			if (r.message) update_charts(r.message);
		},
	});
}

/*── 0.3 chart render (unchanged) ─────────────────────────────────────────*/
function update_charts(data) {
	if (data.error) {
		console.error("Dashboard error:", data.error);
		frappe.msgprint({ title: __("Dashboard Error"), message: data.error, indicator: "red" });
		return;
	}

	const chartConfigs = [
		{ id: "log-type-count",       dataKey: "logTypeCount",       color: "#007bff" },
		{ id: "logs-by-cohort",       dataKey: "logsByCohort",       color: "#17a2b8" },
		{ id: "logs-by-program",      dataKey: "logsByProgram",      color: "#28a745" },
		{ id: "logs-by-author",       dataKey: "logsByAuthor",       color: "#ffc107" },
		{ id: "next-step-types",      dataKey: "nextStepTypes",      color: "#fd7e14" },
		{ id: "incidents-over-time",  dataKey: "incidentsOverTime",  color: "#dc3545", type: "line" },
	];

	chartConfigs.forEach(cfg => {
		const rows = (data[cfg.dataKey] || []).filter(r => r && r.value != null && !isNaN(Number(r.value)));
		if (!rows.length) { $(`#chart-${cfg.id}`).empty(); return; }

		const values = rows.map(r => Number(r.value));
		const total  = values.reduce((a, b) => a + b, 0);
		const labels = rows.map((r, i) => (cfg.type || "bar") === "line"
			? r.label
			: `${r.label} (${total ? Math.round((values[i] / total) * 1000) / 10 : 0}%)`
		);

		new frappe.Chart(`#chart-${cfg.id}`, {
			data: { labels, datasets: [{ values }] },
			type: cfg.type || "bar",
			height: 300,
			colors: [cfg.color],
		});
	});

	if (Array.isArray(data.studentLogs)) {
		const rows = data.studentLogs.map(l => `
			<tr>
				<td>${l.date}</td>
				<td>${l.log_type}</td>
				<td>${l.content}</td>
				<td>${l.author}</td>
			</tr>`).join("");
		document.getElementById("student-log-table-body").innerHTML = rows;
	}
}


frappe.pages["student-log-dashboard"].on_page_load = function (wrapper) {
	frappe.require("/assets/ifitwala_ed/css/dashboard_cards.css");

	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Student Log Dashboard",
		single_column: true,
	});

	const user_default_school = frappe.defaults.get_user_default("school");

	/*── 2. FILTER FIELDS ───────────────────────────────────────────*/
	const school_field = page.add_field({
		fieldname: "school",
		label: __("School"),
		fieldtype: "Link",
		options: "School",
		default: user_default_school,
		change: async () => {
			loadBranchSchools();
			program_field.set_value("");
			selected_student = null;
			studentInput.value = "";
			recent_start = 0;
			fetch_recent_logs(page);
			fetch_dashboard_data(page);
		},
	});
	school_field.set_value(user_default_school);

	async function loadBranchSchools() {
		const root = school_field.get_value();
		if (!root) { branchSchools = []; return; }
		branchSchools = [root];
		try {
			const { message: lr } = await frappe.db.get_value('School', root, ['lft','rgt']);
			if (!lr) return;
			const rows = await frappe.db.get_list('School', {
				fields: ['name'],
				filters: [['lft','>=', lr.lft], ['rgt','<=', lr.rgt]],
				limit: 1000,
			});
			branchSchools = rows.map(r => r.name);
		} catch (e) { console.error('Failed to load branch schools', e); }
	}
	loadBranchSchools();

	const program_field = page.add_field({
		fieldname: "program",
		label: __("Program"),
		fieldtype: "Link",
		options: "Program",
		get_query: () => {
			const sch = school_field.get_value();
			if (sch && branchSchools.length) return { filters: { school: ["in", branchSchools] } };
			if (sch) return { filters: { school: sch } };
			return {};
		},
		change: () => {
			selected_student = null;
			studentInput.value = "";
			recent_start = 0;
			fetch_recent_logs(page);
			fetch_dashboard_data(page);
		},
	});

	const academic_year_field = page.add_field({
		fieldname: "academic_year",
		label: __("Academic Year"),
		fieldtype: "Link",
		options: "Academic Year",
		change: () => {
			recent_start = 0;
			fetch_recent_logs(page);
			fetch_dashboard_data(page);
		}
	});

	const author_field = page.add_field({
		fieldname: "author",
		label: __("Author"),
		fieldtype: "Link",
		options: "Employee",
		change: () => { recent_start = 0; fetch_recent_logs(page); fetch_dashboard_data(page); },
	});

	// ★ NEW: From / To Date filters
	const from_date_field = page.add_field({
		fieldname: "from_date",
		label: __("From Date"),
		fieldtype: "Date",
		change: () => { recent_start = 0; fetch_recent_logs(page); fetch_dashboard_data(page); },
	});
	const to_date_field = page.add_field({
		fieldname: "to_date",
		label: __("To Date"),
		fieldtype: "Date",
		change: () => { recent_start = 0; fetch_recent_logs(page); fetch_dashboard_data(page); },
	});

	/*── 3. MAIN CONTENT HTML ───────────────────────────────────────────────*/
	$(wrapper).append(`
		<div class="dashboard-overlay" id="dashboard-overlay"></div>

		<div class="dashboard-content container">

			${createDashboardCard("incidents-over-time", "Incidents Over Time")}
			${recentLogsCard()}
			${createDashboardCard("log-type-count",  "Log Type Count")}
			${createDashboardCard("logs-by-cohort",  "Logs by Cohort")}
			${createDashboardCard("logs-by-program", "Logs by Program")}
			${createDashboardCard("logs-by-author",  "Logs by Author")}
			${createDashboardCard("next-step-types", "Next Step Types")}

			${studentLogDetailCard()}

		</div>
	`);

	function createDashboardCard(id, title) {
		return `
			<div class="dashboard-card" id="${id}">
				<div class="card-title">${title}</div>
				<div id="chart-${id}"></div>
			</div>`;
	}

	function recentLogsCard() {
		return `
			<div class="dashboard-card" id="recent-log-card">
				<div class="card-title">Recent Student Logs</div>

				<div class="student-log-table-wrapper">
					<table class="table table-bordered table-hover">
						<thead class="table-light">
							<tr>
								<th>Date</th><th>Student</th><th>Program</th><th>Log Type</th>
								<th>Log</th><th>Author</th><th>⧈</th>
							</tr>
						</thead>
						<tbody id="recent-log-table-body"></tbody>
					</table>
				</div>

				<div class="load-more-wrapper">
					<button class="btn btn-link w-100" id="recent-log-load-more">
						Load more…
					</button>
				</div>
			</div>`;
	}

	function studentLogDetailCard() {
		return `
			<div class="dashboard-card" id="student-log-detail">
				<div class="card-title">Student Log Detail</div>

				<div class="student-log-filter mb-2">
					<label class="form-label" for="student-select">Student</label>
					<input class="form-control" type="text" id="student-select"
						   placeholder="Start typing…" autocomplete="off">
					<div id="student-dropdown" class="student-dropdown"></div>
				</div>

				<div class="student-log-table-wrapper">
					<table class="table table-bordered table-hover student-log-table">
						<thead class="table-light">
							<tr>
								<th>Date</th><th>Log Type</th>
								<th>Log Content</th><th>Author</th>
							</tr>
						</thead>
						<tbody id="student-log-table-body"></tbody>
					</table>
				</div>
			</div>`;
	}

	document.getElementById("recent-log-load-more")
		.addEventListener("click", () => fetch_recent_logs(page, true));

	/*── 5. STUDENT FILTER LOGIC (unchanged) ───────────────────────────────*/
	const studentInput = document.getElementById("student-select");
	const dropdownEl   = document.getElementById("student-dropdown");

	studentInput.addEventListener("input", () => {
		const query = studentInput.value.trim();
		if (!query) return hideStudentSuggestions();

		const filters = {
			school: page.fields_dict.school.get_value(),
			program: page.fields_dict.program.get_value(),
			academic_year: page.fields_dict.academic_year.get_value(),
		};

		frappe.call({
			method: "ifitwala_ed.students.page.student_log_dashboard.student_log_dashboard.get_distinct_students",
			args: { filters: JSON.stringify(filters), search_text: query },
			callback: (r) => {
				if (r.message && !r.message.error) {
					const suggestions = r.message.map(s => `
						<div class="student-suggestion"
							 data-id="${s.student}"
							 data-name="${s.student_full_name}">
							${s.student_full_name} (${s.student})
						</div>`).join("");
					showStudentSuggestions(suggestions);
				}
			},
		});
	});

	studentInput.addEventListener("keydown", (e) => {
		if (e.key === "Enter") {
			e.preventDefault();
			e.stopPropagation();
			const first = dropdownEl.querySelector(".student-suggestion");
			if (first) first.click();
		}
	});

	dropdownEl.addEventListener("click", (e) => {
		e.stopPropagation();
		const target = e.target.closest(".student-suggestion");
		if (!target) return;
		selected_student   = target.dataset.id;
		studentInput.value = target.dataset.name;
		hideStudentSuggestions();
		fetch_dashboard_data(page);
	});

	document.addEventListener("click", (e) => {
		if (e.target !== studentInput && !dropdownEl.contains(e.target)) {
			hideStudentSuggestions();
		}
	});

	function showStudentSuggestions(html) {
		dropdownEl.innerHTML = html;
		dropdownEl.style.display = "block";
	}
	function hideStudentSuggestions() {
		dropdownEl.innerHTML = "";
		dropdownEl.style.display = "none";
	}

	/*── 6. ZOOM UX (unchanged) ───────────────────────────────────────────*/
	function toggleZoom(card) {
		const overlay = document.getElementById("dashboard-overlay");
		if (card.classList.contains("zoomed")) {
			card.classList.remove("zoomed");
			overlay.classList.remove("active");
		} else {
			document.querySelectorAll(".dashboard-card.zoomed").forEach(c => {
				if (c !== card) c.classList.remove("zoomed");
			});
			card.classList.add("zoomed");
			overlay.classList.add("active");
		}
	}
	document.querySelectorAll(".dashboard-card").forEach(card => {
		card.addEventListener("click", (e) => {
			if (card.id === "student-log-detail" &&
				e.target.closest(".student-log-filter, .student-dropdown")) {
				e.stopPropagation();
				return;
			}
			toggleZoom(card);
		});
	});
	document.getElementById("dashboard-overlay").addEventListener("click", () => {
		document.querySelectorAll(".dashboard-card.zoomed").forEach(c => c.classList.remove("zoomed"));
		document.getElementById("dashboard-overlay").classList.remove("active");
	});

	/*── 7. INITIAL LOAD ───────────────────────────────────────────────────*/
	fetch_recent_logs(page);
	fetch_dashboard_data(page);
};

File: ifitwala_ed/students/page/student_log_dashboard/student_log_dashboard.py
# Copyright (c) 2025, Francois de Ryckel and contributors
# For license information, please see license.txt

"""Uses consistent alias `sl` for tabStudent Log in every query to
   avoid ambiguous column references when filters include `program`.
"""

import frappe


@frappe.whitelist()
def get_dashboard_data(filters=None):
	try:
		filters = frappe.parse_json(filters) or {}

		authorized_schools = get_authorized_schools(frappe.session.user)
		if not authorized_schools:
			return {"error": "No authorized schools found."}

		conditions, params = [], {}
		# ▶ enforce access by school (uses your new index on sl.school)
		conditions.append("sl.school IN %(authorized_schools)s")
		params["authorized_schools"] = tuple(authorized_schools)

		# ── School filter (now direct on Student Log) ─────────────
		if filters.get("school"):
			conditions.append("sl.school = %(field_school)s")
			params["field_school"] = filters["school"]

		# ── Direct columns in Student Log (alias sl) ──────────────
		direct_map = {
			"academic_year": "sl.academic_year",
			"program":       "sl.program",
			"author":        "sl.author_name",
		}
		for key, col in direct_map.items():
			if filters.get(key):
				ph = f"field_{key}"
				conditions.append(f"{col} = %({ph})s")
				params[ph] = filters[key]

		where_clause = " AND ".join(conditions) if conditions else "1=1"

		def q(sql):
			return frappe.db.sql(sql.format(w=where_clause), params, as_dict=True)

		# (no query bodies change below — they already read from sl.*)
		log_type_count = q(
			"SELECT sl.log_type AS label, COUNT(*) AS value "
			"FROM `tabStudent Log` sl WHERE {w} "
			"GROUP BY sl.log_type ORDER BY value DESC"
		)

		logs_by_cohort = q(
			"SELECT pe.cohort AS label, COUNT(sl.name) AS value "
			"FROM `tabStudent Log` sl "
			"LEFT JOIN `tabProgram Enrollment` pe ON sl.student = pe.student "
			"WHERE {w} GROUP BY pe.cohort ORDER BY value DESC"
		)

		logs_by_program = q(
			"SELECT sl.program AS label, COUNT(sl.name) AS value "
			"FROM `tabStudent Log` sl WHERE {w} GROUP BY sl.program ORDER BY value DESC"
		)

		logs_by_author = q(
			"SELECT sl.author_name AS label, COUNT(*) AS value "
			"FROM `tabStudent Log` sl WHERE {w} GROUP BY sl.author_name ORDER BY value DESC"
		)

		next_step_types = q(
			"SELECT sl.next_step AS label, COUNT(*) AS value "
			"FROM `tabStudent Log` sl WHERE {w} GROUP BY sl.next_step ORDER BY value DESC"
		)

		incidents_over_time = q(
			"SELECT DATE_FORMAT(sl.date,'%%Y-%%m-%%d') AS label, COUNT(*) AS value "
			"FROM `tabStudent Log` sl WHERE {w} GROUP BY label ORDER BY label ASC"
		)

		open_follow_ups = frappe.db.sql(
			f"SELECT COUNT(*) FROM `tabStudent Log` sl WHERE {where_clause} "
			"AND sl.follow_up_status = 'Open'",
			params,
		)[0][0]

		# ── Student Logs (if student filter is set) ─────────────────
		student_logs = []
		if filters.get("student"):
				# reuse the same params + conditions used above
				conds = ["sl.school IN %(authorized_schools)s", "sl.student = %(field_student)s"]
				p = {**params, "field_student": filters["student"]}

				# if the page has a School filter set, honor it too (cheap on sl.school)
				if filters.get("school"):
						conds.append("sl.school = %(field_school)s")
						p["field_school"] = filters["school"]

				where_detail = " AND ".join(conds)

				student_logs = frappe.db.sql(
						f"""
						SELECT
								sl.date,
								sl.log_type,
								sl.log AS content,
								sl.author_name AS author
						FROM `tabStudent Log` sl
						WHERE {where_detail}
						ORDER BY sl.date DESC
						""",
						p,
						as_dict=True
				)

		return {
			"logTypeCount": log_type_count,
			"logsByCohort": logs_by_cohort,
			"logsByProgram": logs_by_program,
			"logsByAuthor": logs_by_author,
			"nextStepTypes": next_step_types,
			"incidentsOverTime": incidents_over_time,
			"openFollowUps": open_follow_ups,
			"studentLogs": student_logs,
		}

	except Exception as e:
		frappe.log_error(str(e), "Student Log Dashboard Data Error")
		return {"error": str(e)}


@frappe.whitelist()
def get_distinct_students(filters=None, search_text: str = ""):
	try:
		filters = frappe.parse_json(filters) or {}
		txt = (search_text or "").strip()

		# Authorize by user's school branch (self + descendants)
		authorized_schools = get_authorized_schools(frappe.session.user)
		if not authorized_schools:
			return {"error": "No authorized schools found."}

		conditions, params = [], {}
		params["authorized_schools"] = tuple(authorized_schools)

		# Context filters
		if filters.get("school"):
			conditions.append("pe.school = %(school)s")
			params["school"] = filters["school"]
		if filters.get("program"):
			conditions.append("pe.program = %(program)s")
			params["program"] = filters["program"]
		if filters.get("academic_year"):
			conditions.append("pe.academic_year = %(academic_year)s")
			params["academic_year"] = filters["academic_year"]

		# Partial text match (id or full name)
		if txt:
			conditions.append("(pe.student LIKE %(txt)s OR s.student_full_name LIKE %(txt)s)")
			params["txt"] = f"%{txt}%"

		# 🔒 Always constrain to authorized schools (no siblings)
		conditions.append("pe.school IN %(authorized_schools)s")

		where_clause = " AND ".join(conditions) if conditions else "1=1"

		return frappe.db.sql(f"""
			SELECT DISTINCT pe.student, s.student_full_name AS student_full_name
			FROM `tabProgram Enrollment` pe
			INNER JOIN `tabStudent` s ON pe.student = s.name
			WHERE {where_clause}
			ORDER BY s.student_full_name
			LIMIT 100
		""", params, as_dict=True)

	except Exception as e:
		frappe.log_error(message=str(e), title="Student Lookup Error")
		return {"error": str(e)}


@frappe.whitelist()
def get_recent_logs(filters=None, start: int = 0, page_length: int = 25):
	filters = frappe.parse_json(filters) or {}

	authorized_schools = get_authorized_schools(frappe.session.user)
	if not authorized_schools:
		return []

	conditions, params = [], {}

	# 🔒 scope by school branch (self + descendants)
	conditions.append("sl.school IN %(authorized_schools)s")
	params["authorized_schools"] = tuple(authorized_schools)

	# page filters (same as before), but school now direct on Student Log
	if filters.get("school"):
		conditions.append("sl.school = %(school)s")
		params["school"] = filters["school"]

	direct_map = {
		"academic_year": "sl.academic_year",
		"program":       "sl.program",
		"author":        "sl.author_name",
	}
	for key, col in direct_map.items():
		if filters.get(key):
			conditions.append(f"{col} = %({key})s")
			params[key] = filters[key]

	where_clause = " AND ".join(conditions) if conditions else "1=1"

	logs = frappe.db.sql(
		f"""
		SELECT
			sl.date,
			s.student_full_name AS student,
			sl.program,
			sl.log_type,
			sl.log              AS content,
			sl.author_name      AS author,
			sl.requires_follow_up
		FROM `tabStudent Log` sl
		LEFT JOIN `tabStudent` s ON s.name = sl.student
		WHERE {where_clause}
		ORDER BY sl.date DESC
		LIMIT %(start)s, %(page_len)s
		""",
		{**params, "start": start, "page_len": page_length},
		as_dict=True,
	)
	return logs


def get_authorized_schools(user):
	"""Return user's school + all descendants using one SQL join on lft/rgt."""
	# Resolve the anchor school: user default first, else Employee.school
	default_school = frappe.defaults.get_user_default("school", user)
	if not default_school:
		default_school = frappe.db.get_value("Employee", {"user_id": user}, "school")
	if not default_school:
		return []

	rows = frappe.db.sql(
		"""
		SELECT s2.name
		FROM `tabSchool` s1
		JOIN `tabSchool` s2
			ON s2.lft >= s1.lft AND s2.rgt <= s1.rgt
		WHERE s1.name = %s
		""",
		(default_school,),
		as_list=True,
	)
	return [r[0] for r in rows] or [default_school]

